extensions/ffi/lib.rs

```diff
@@ -0,0 +1,192 @@
+      FFIType::I32 => libffi::middle::Arg::new(&(arg.as_i64() as i32)),
+      FFIType::U64 => libffi::middle::Arg::new(&arg.as_u64()),
+      FFIType::I64 => libffi::middle::Arg::new(&arg.as_i64()),
+      FFIType::USize => libffi::middle::Arg::new(&(arg.as_u64() as usize)),
+      FFIType::ISize => libffi::middle::Arg::new(&(arg.as_i64() as isize)),
+      FFIType::F32 => libffi::middle::Arg::new(&(arg.as_f64() as f32)),
+      FFIType::F64 => libffi::middle::Arg::new(&arg.as_f64()),
+    }
+  }
+}
+
+impl FFIArg {
+  fn as_u64(self) -> u64 {
+    self
+      .value
+      .as_u64()
+      .expect("Expected ffi arg value to be an unsigned integer")
+  }
+
+  fn as_i64(self) -> i64 {
+    self
+      .value
+      .as_i64()
+      .expect("Expected ffi arg value to be a signed integer")
+  }
+
+  fn as_f64(self) -> f64 {
+    self
+      .value
+      .as_f64()
+      .expect("Expected ffi arg value to be a float")
+  }
+}
+
+#[derive(Deserialize, Clone, Copy)]
+#[serde(rename_all = "lowercase")]
+enum FFIType {
+  Void,
+  U8,
+  I8,
+  U16,
+  I16,
+  U32,
+  I32,
+  U64,
+  I64,
+  USize,
+  ISize,
+  F32,
+  F64,
+  //  Ptr,
+  //  CStr,
+  //  Struct(Vec<FFIType>),
+}
+
+impl From<FFIType> for libffi::middle::Type {
+  fn from(r#type: FFIType) -> Self {
+    match r#type {
+      FFIType::Void => libffi::middle::Type::void(),
+      FFIType::U8 => libffi::middle::Type::u8(),
+      FFIType::I8 => libffi::middle::Type::i8(),
+      FFIType::U16 => libffi::middle::Type::u16(),
+      FFIType::I16 => libffi::middle::Type::i16(),
+      FFIType::U32 => libffi::middle::Type::u32(),
+      FFIType::I32 => libffi::middle::Type::i32(),
+      FFIType::U64 => libffi::middle::Type::u64(),
+      FFIType::I64 => libffi::middle::Type::i64(),
+      FFIType::USize => libffi::middle::Type::usize(),
+      FFIType::ISize => libffi::middle::Type::isize(),
+      FFIType::F32 => libffi::middle::Type::f32(),
+      FFIType::F64 => libffi::middle::Type::f64(),
+    }
+  }
+}
+
+#[derive(Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct DlcallArgs {
+  rid: ResourceId,
+  sym: String,
+  args: Vec<FFIArg>,
+  return_type: FFIType,
+}
+
+fn op_dlcall(
+  state: &mut deno_core::OpState,
+  dlcall_args: DlcallArgs,
+  _: (),
+) -> Result<Value, AnyError> {
+  let library = state
+    .resource_table
+    .get::<LibraryResource>(dlcall_args.rid)
+    .ok_or_else(bad_resource_id)?;
+  let fn_ptr = unsafe { library.0.symbol::<*const c_void>(&dlcall_args.sym) }?;
+  let fn_code_ptr = libffi::middle::CodePtr::from_ptr(fn_ptr as _);
+  let types = dlcall_args
+    .args
+    .clone()
+    .into_iter()
+    .map(|arg| arg.ffi_type.into());
+  let cif = Cif::new(types, dlcall_args.return_type.into());
+  let args: Vec<libffi::middle::Arg> =
+    dlcall_args.args.into_iter().map(|arg| arg.into()).collect();
+
+  Ok(match dlcall_args.return_type {
+    FFIType::Void => json!(unsafe { cif.call::<()>(fn_code_ptr, &args) }),
```

Comment:
Is `serde_v8` a better option here?

Reply:
the `json!` macro automatically makes it into a `serde_v8` serializable `Value` if I'm not mistaken